# Predicting Compresive Strenght of Concrete Using Machine Learning
Compressive strength is the most widely recognized and accepted measure of concrete's strength. It serves as the primary criterion for assessing whether a specific concrete mixture can endure the structural loads it will encounter. Often referred to as the "nameplate" rating, compressive strength is the most frequently cited attribute in construction specifications.

To test compressive strength, cylindrical concrete samples are broken in a specialized machine, following the ASTM (American Society for Testing & Materials) standard C39 protocol.

Compressive strength of concrete is measured in pounds per square inch (psi), representing the force in pounds applied over a square inch area, or in megapascals (MPa) for metric units. This measurement is obtained by applying force from opposite sides of the concrete sample until it fractures, revealing the strength limits of the cured mix. Aggregate materials within the concrete help distribute and counterbalance the load. Generally, a higher psi indicates greater compressive strength and typically comes with a higher cost, but it also implies enhanced durability, longevity, and often a more efficient use of material compared to lower-strength mixes.

Compressive strength is normally tested at seven days and then again at 28 days. The seven-day test determines early strength gains and verifies that the mix is on track to set properly. The final cured design strength (and the basis for minimum design values) is the 28-day test as noted in the ACI standards.

In this project,  we will use the following features: 
- **Cement**: kg in a m³ mixture
- **Blast Furnace Slag**: kg in a m³ mixture
- **Fly Ash**: kg in a m³ mixture
- **Water**: kg in a m³ mixture
- **Superplasticizer**: kg in a m³ mixture
- **Coarse Aggregate**: kg in a m³ mixture
- **Fine Aggregate**: kg in a m³ mixture
- **Age**: Day (1~365)

to predict the target variable: 
- **Concrete Compressive Strength**: MPa



For this project we will use Mean Squared Error (MSE) for both cost function and evaluation metrics. 

In [140]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score

from sklearn.linear_model import SGDRegressor
from sklearn.linear_model import Lasso
from sklearn.linear_model import Ridge

df = pd.read_csv('/kaggle/input/regression-with-neural-networking/concrete_data.csv')
df2 = pd.read_csv('/kaggle/input/regression-with-neural-networking/concrete_data.csv')
df3 = pd.read_csv('/kaggle/input/regression-with-neural-networking/concrete_data.csv')



In [141]:
df

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0      540.0                 0.0      0.0  162.0               2.5   
1      540.0                 0.0      0.0  162.0               2.5   
2      332.5               142.5      0.0  228.0               0.0   
3      332.5               142.5      0.0  228.0               0.0   
4      198.6               132.4      0.0  192.0               0.0   
...      ...                 ...      ...    ...               ...   
1025   276.4               116.0     90.3  179.6               8.9   
1026   322.2                 0.0    115.6  196.0              10.4   
1027   148.5               139.4    108.6  192.7               6.1   
1028   159.1               186.7      0.0  175.6              11.3   
1029   260.9               100.5     78.3  200.6               8.6   

      Coarse Aggregate  Fine Aggregate  Age  Strength  
0               1040.0           676.0   28     79.99  
1               1055.0           676.0   28     61.89  
2                932.0           594.0  270     40.27  
3                932.0           594.0  365     41.05  
4                978.4           825.5  360     44.30  
...                ...             ...  ...       ...  
1025             870.1           768.3   28     44.28  
1026             817.9           813.4   28     31.18  
1027             892.4           780.0   28     23.70  
1028             989.6           788.9   28     32.77  
1029             864.5           761.5   28     32.40  

[1030 rows x 9 columns]

By printing the information about the dataframe, we have foundthat there is no missing values and each variables has the correct datatype. 

In [142]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1030 entries, 0 to 1029
Data columns (total 9 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Cement              1030 non-null   float64
 1   Blast Furnace Slag  1030 non-null   float64
 2   Fly Ash             1030 non-null   float64
 3   Water               1030 non-null   float64
 4   Superplasticizer    1030 non-null   float64
 5   Coarse Aggregate    1030 non-null   float64
 6   Fine Aggregate      1030 non-null   float64
 7   Age                 1030 non-null   int64  
 8   Strength            1030 non-null   float64
dtypes: float64(8), int64(1)
memory usage: 72.5 KB


## Regression using SGDREgressor Class
SGD stands for Stochastic Gradient Descent: the gradient of the loss is estimated each sample at a time and the model is updated along the way with a decreasing strength schedule (aka learning rate).

We will use the SGDRegressor class from sklearn.linear_model package.Prior to this, we need to scale the features since the SGDRegressor class is sensitive to the scale of the features. Large difference in features scales can lead to numerical instability during gradient computation. 

#### Extracting the Features & Splitting

In [143]:
#extract the explanatory and response variable
#axis = 1 indicates that we are droping a column
#axis = 1 indicates that we are droping from row
X = df.drop('Strength', axis = 1)
Y = df['Strength']

#split into train and test. use 80-20 rule
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2, random_state = 42)

#### Training
In this part, the loss='squared_error' argument indicates that the cost function to be used is MSE, the learning rate (alpha) is set to 0.0001, and the tolerance is e^-3. This indicates that convergence is achieved if the cost function decreases by <= tolerance. The model does not necessarily need to repeat 1000 times if it converges sooner. 

In [144]:
from sklearn.preprocessing import StandardScaler

#instance
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.fit_transform(X_test)

#instance of the model and training
model = SGDRegressor(loss='squared_error', alpha=0.00001, max_iter=1000, tol=1e-3, random_state=42, verbose = 1)
model.fit(X_train_scaled, Y_train)


-- Epoch 1
Norm: 10.97, NNZs: 8, Bias: 31.470546, T: 824, Avg. loss: 190.203997
Total training time: 0.00 seconds.
-- Epoch 2
Norm: 12.10, NNZs: 8, Bias: 34.636910, T: 1648, Avg. loss: 63.275791
Total training time: 0.00 seconds.
-- Epoch 3
Norm: 12.64, NNZs: 8, Bias: 35.385086, T: 2472, Avg. loss: 59.210603
Total training time: 0.00 seconds.
-- Epoch 4
Norm: 12.91, NNZs: 8, Bias: 35.794916, T: 3296, Avg. loss: 58.434260
Total training time: 0.00 seconds.
-- Epoch 5
Norm: 13.16, NNZs: 8, Bias: 35.821443, T: 4120, Avg. loss: 58.079001
Total training time: 0.00 seconds.
-- Epoch 6
Norm: 13.41, NNZs: 8, Bias: 35.739610, T: 4944, Avg. loss: 57.734515
Total training time: 0.00 seconds.
-- Epoch 7
Norm: 13.65, NNZs: 8, Bias: 35.853215, T: 5768, Avg. loss: 57.457101
Total training time: 0.00 seconds.
-- Epoch 8
Norm: 13.77, NNZs: 8, Bias: 35.760282, T: 6592, Avg. loss: 57.472421
Total training time: 0.00 seconds.
-- Epoch 9
Norm: 13.84, NNZs: 8, Bias: 35.770327, T: 7416, Avg. loss: 57.330591


SGDRegressor(alpha=1e-05, random_state=42, verbose=1)

#### Evaluation


In [145]:
#evaluating the model's performance
Y_pred = model.predict(X_test_scaled)
mse = mean_squared_error(Y_test, Y_pred)
print(f"Mean squared error on testing set: {mse}")

mpe = np.mean((Y_pred - Y_test) / Y_test) * 100
print(f"Mean percentage error on testing set: {mpe} %")

Mean squared error on testing set: 97.48045906978886
Mean percentage error on testing set: 13.323840623231206 %


The model converges after 35th epoch (iteration). From our evaluation metrics, the MSE is 97.480, which is equivalent to 13.324% in Mean Percentage Error (MPE). MPE focuses on the percentage difference between predicted and actual values. We can improve this model by including engineered features. 

# Feature Engineering - Manually

Feature engineering is the process of selecting, manipulating and transforming raw data (or features) into features that can be used in supervised learning. This can involved tranforming the existing feature (e.g., taking the log) or combining features. 

In [146]:

df['Aggregate Content'] = df['Coarse Aggregate'] + df['Fine Aggregate']

# Total Binder Content
df['Total Binder Content'] = df['Cement'] + df['Blast Furnace Slag'] + df['Fly Ash']

# Water to Cement Ratio (W/C Ratio)
df['Water to Cement Ratio'] = df['Water'] / df['Cement']

# Water to Binder Ratio (W/B Ratio)
df['Water to Binder Ratio'] = df['Water'] / df['Total Binder Content']

# Superplasticizer to Binder Ratio
df['Superplasticizer to Binder Ratio'] = df['Superplasticizer'] / df['Total Binder Content']

# Aggregate Cggregate Ratio
df['Fine to Total Aggregate Ratio'] = df['Fine Aggregate'] / df['Aggregate Content']

# Binder to Aggregate Ratio
df['Binder to Aggregate Ratio'] = df['Total Binder Content'] / df['Aggregate Content']

# Cement Proportion in Binder
df['Cement Proportion in Binder'] = df['Cement'] / df['Total Binder Content']

# Slag Proportion in Binder
df['Slag Proportion in Binder'] = df['Blast Furnace Slag'] / df['Total Binder Content']

# Fly Ash Proportion in Binder
df['Fly Ash Proportion in Binder'] = df['Fly Ash'] / df['Total Binder Content']

# Water Content per Unit Binder
df['Water Content per Unit Binder'] = df['Water'] / df['Total Binder Content']

# Superplasticizer Content per Unit Binder
df['Superplasticizer Content per Unit Binder'] = df['Superplasticizer'] / df['Total Binder Content']

# Total Aggregate to Cement Ratio
df['Total Aggregate to Cement Ratio'] = df['Aggregate Content'] / df['Cement']

# Age Squared
df['Age Squared'] = df['Age'] ** 2




In [147]:
df

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0      540.0                 0.0      0.0  162.0               2.5   
1      540.0                 0.0      0.0  162.0               2.5   
2      332.5               142.5      0.0  228.0               0.0   
3      332.5               142.5      0.0  228.0               0.0   
4      198.6               132.4      0.0  192.0               0.0   
...      ...                 ...      ...    ...               ...   
1025   276.4               116.0     90.3  179.6               8.9   
1026   322.2                 0.0    115.6  196.0              10.4   
1027   148.5               139.4    108.6  192.7               6.1   
1028   159.1               186.7      0.0  175.6              11.3   
1029   260.9               100.5     78.3  200.6               8.6   

      Coarse Aggregate  Fine Aggregate  Age  Strength  Aggregate Content  ...  \
0               1040.0           676.0   28     79.99             1716.0  ...   
1               1055.0           676.0   28     61.89             1731.0  ...   
2                932.0           594.0  270     40.27             1526.0  ...   
3                932.0           594.0  365     41.05             1526.0  ...   
4                978.4           825.5  360     44.30             1803.9  ...   
...                ...             ...  ...       ...                ...  ...   
1025             870.1           768.3   28     44.28             1638.4  ...   
1026             817.9           813.4   28     31.18             1631.3  ...   
1027             892.4           780.0   28     23.70             1672.4  ...   
1028             989.6           788.9   28     32.77             1778.5  ...   
1029             864.5           761.5   28     32.40             1626.0  ...   

      Superplasticizer to Binder Ratio  Fine to Total Aggregate Ratio  \
0                             0.004630                       0.393939   
1                             0.004630                       0.390526   
2                             0.000000                       0.389253   
3                             0.000000                       0.389253   
4                             0.000000                       0.457620   
...                                ...                            ...   
1025                          0.018438                       0.468933   
1026                          0.023755                       0.498621   
1027                          0.015385                       0.466396   
1028                          0.032678                       0.443576   
1029                          0.019559                       0.468327   

      Binder to Aggregate Ratio  Cement Proportion in Binder  \
0                      0.314685                     1.000000   
1                      0.311958                     1.000000   
2                      0.311271                     0.700000   
3                      0.311271                     0.700000   
4                      0.183491                     0.600000   
...                         ...                          ...   
1025                   0.294617                     0.572612   
1026                   0.268375                     0.735952   
1027                   0.237084                     0.374527   
1028                   0.194434                     0.460093   
1029                   0.270418                     0.593359   

      Slag Proportion in Binder  Fly Ash Proportion in Binder  \
0                      0.000000                      0.000000   
1                      0.000000                      0.000000   
2                      0.300000                      0.000000   
3                      0.300000                      0.000000   
4                      0.400000                      0.000000   
...                         ...                           ...   
1025                   0.240315                      0.187073   
1026                   0.0

#### Extracting the Features & Splitting

In [148]:
#extract the explanatory and response variable
#axis = 1 indicates that we are droping a column
#axis = 1 indicates that we are droping from row
X = df.drop('Strength', axis = 1)
Y = df['Strength']

#split into train and test. use 80-20 rule
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2, random_state = 42)

#### Training

In [149]:
from sklearn.preprocessing import StandardScaler

#instance
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.fit_transform(X_test)

#instance of the model and training
model = SGDRegressor(loss='squared_error', alpha=0.00001, max_iter=1000, tol=1e-3, random_state=42, verbose = 1)
model.fit(X_train_scaled, Y_train)


-- Epoch 1
Norm: 7.08, NNZs: 22, Bias: 31.480429, T: 824, Avg. loss: 184.616563
Total training time: 0.00 seconds.
-- Epoch 2
Norm: 8.40, NNZs: 22, Bias: 34.656460, T: 1648, Avg. loss: 59.445908
Total training time: 0.00 seconds.
-- Epoch 3
Norm: 9.96, NNZs: 22, Bias: 35.383297, T: 2472, Avg. loss: 53.014941
Total training time: 0.00 seconds.
-- Epoch 4
Norm: 11.16, NNZs: 22, Bias: 35.767834, T: 3296, Avg. loss: 49.966974
Total training time: 0.00 seconds.
-- Epoch 5
Norm: 12.36, NNZs: 22, Bias: 35.811245, T: 4120, Avg. loss: 47.967444
Total training time: 0.00 seconds.
-- Epoch 6
Norm: 13.44, NNZs: 22, Bias: 35.698386, T: 4944, Avg. loss: 46.076160
Total training time: 0.00 seconds.
-- Epoch 7
Norm: 14.48, NNZs: 22, Bias: 35.868651, T: 5768, Avg. loss: 44.741301
Total training time: 0.00 seconds.
-- Epoch 8
Norm: 15.24, NNZs: 22, Bias: 35.777716, T: 6592, Avg. loss: 43.798907
Total training time: 0.00 seconds.
-- Epoch 9
Norm: 16.00, NNZs: 22, Bias: 35.807683, T: 7416, Avg. loss: 42.8

SGDRegressor(alpha=1e-05, random_state=42, verbose=1)

In [150]:
#evaluating the model's performance
Y_pred = model.predict(X_test_scaled)
mse = mean_squared_error(Y_test, Y_pred)
print(f"Mean squared error on testing set: {mse}")

mpe = np.mean((Y_pred - Y_test) / Y_test) * 100
print(f"Mean percentage error on testing set: {mpe} %")

Mean squared error on testing set: 63.07437383378309
Mean percentage error on testing set: 9.459451079616626 %


Observe that the cost function is reduced. The prior is 97.480, and after including the engineered features the cost function is reduced to 63.67. However, it look longer for the model to converge. This is expected since there are more features. 

## Polynomial Regression using LinearRegression() class
This class implements linear regression models, from simple linear regression to multiple and polynomial regression. This uses the Ordinary Least Squares (OLS) algorithm to fit a linear regression model. This algorithm finds the coefficients of the linear equation that minimizes the squared differences betweent the observed and predicted values of the target variable. 

### Generating Polynomial Features using PolynomialFeature Class from scikitlearn.preprocessing

PolynomialFeature class generate a new feature matrix consisting of all polynomial combinations of the features with degree less than or equal to the specified degree. For example, if an input sample is two dimensional and of the form [a, b], the degree-2 polynomial features are [1, a, b, a^2, ab, b^2].

In [151]:
#extract the explanatory and response variable
#axis = 1 indicates that we are droping a column
#axis = 1 indicates that we are droping from row
X = df2.drop('Strength', axis = 1)
Y = df2['Strength']

#split into train and test. use 80-20 rule
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2, random_state = 42)

In [152]:
from sklearn.preprocessing import PolynomialFeatures

#instance 
poly_features = PolynomialFeatures(degree = 2, include_bias = False)



#create poly features for X_trian and X_test
X_train_poly = poly_features.fit_transform(X_train)
X_test_poly = poly_features.fit_transform(X_test)

In [153]:

# Create and fit the model
model = LinearRegression()
model.fit(X_train_poly, Y_train)

# Make predictions
Y_test_pred = model.predict(X_test_poly)


#evaluating the model's performance
Y_pred = model.predict(X_test_poly)
mse = mean_squared_error(Y_test, Y_pred)
print(f"Mean squared error on testing set: {mse}")

mpe = np.mean((Y_pred - Y_test) / Y_test) * 100
print(f"Mean percentage error on testing set: {mpe} %")

Mean squared error on testing set: 55.582457881024105
Mean percentage error on testing set: 5.535528217238074 %


**By including the polynomial terms and interaction terms in our model, 
the mpe is reduced from 9.45 (This include engineered features based on material science) % to 1.14% (while this include all combination, i.e., All polynomial combination
of features with degree less than or euqal to the specified degree).** 

Using the original features to create new polynomial features and interaction features using PolynomialFeature class, we can see a significant reduction to the cost function (MSE = 55.582). This model is better compared to the preceeding models. 

## Regularized Linear Regression
Regularized regression is a technique used in machine learning to improve the performance and generalizability of regression models by adding a penalty term to the loss function. This penalty discourages complex models that are likely to overfit the training data. Regularization methods adjust the model to be simpler, often resulting in better performance on unseen data.

Types include:
*     -Ridge Regression (L2 Regularization)
*     -Lasso Regression (L1 Regulzarization)


### L1 Regularization

In [154]:
#extract the explanatory and response variable
#axis = 1 indicates that we are droping a column
#axis = 1 indicates that we are droping from row
X = df3.drop('Strength', axis = 1)
Y = df3['Strength']

#split into train and test. use 80-20 rule
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2, random_state = 42)

In [155]:
df3

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0      540.0                 0.0      0.0  162.0               2.5   
1      540.0                 0.0      0.0  162.0               2.5   
2      332.5               142.5      0.0  228.0               0.0   
3      332.5               142.5      0.0  228.0               0.0   
4      198.6               132.4      0.0  192.0               0.0   
...      ...                 ...      ...    ...               ...   
1025   276.4               116.0     90.3  179.6               8.9   
1026   322.2                 0.0    115.6  196.0              10.4   
1027   148.5               139.4    108.6  192.7               6.1   
1028   159.1               186.7      0.0  175.6              11.3   
1029   260.9               100.5     78.3  200.6               8.6   

      Coarse Aggregate  Fine Aggregate  Age  Strength  
0               1040.0           676.0   28     79.99  
1               1055.0           676.0   28     61.89  
2                932.0           594.0  270     40.27  
3                932.0           594.0  365     41.05  
4                978.4           825.5  360     44.30  
...                ...             ...  ...       ...  
1025             870.1           768.3   28     44.28  
1026             817.9           813.4   28     31.18  
1027             892.4           780.0   28     23.70  
1028             989.6           788.9   28     32.77  
1029             864.5           761.5   28     32.40  

[1030 rows x 9 columns]

In [156]:
# from sklearn.preprocessing import StandardScaler

# #instance
# scaler = StandardScaler()
# X_train_scaled = scaler.fit_transform(X_train)
# X_test_scaled = scaler.fit_transform(X_test)

# #instance of the model and training
# model = SGDRegressor(loss='squared_error', alpha=0.00001, max_iter=1000, tol=1e-3, random_state=42, verbose = 1)
# model.fit(X_train_scaled, Y_train)


#instance of the model and training
model = Lasso(alpha = 0.1, random_state = 42)
model.fit(X_train, Y_train)


Lasso(alpha=0.1, random_state=42)

In [157]:
#evaluating the model's performance
Y_pred = model.predict(X_test)
mse = mean_squared_error(Y_test, Y_pred)
print(f"Mean squared error on testing set: {mse}")

mpe = np.mean((Y_pred - Y_test) / Y_test) * 100
print(f"Mean percentage error on testing set: {mpe} %")

Mean squared error on testing set: 95.9651693895091
Mean percentage error on testing set: 12.825271750806857 %


In [158]:
from sklearn.preprocessing import StandardScaler

#instance
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.fit_transform(X_test)



#instance of the model and training
model = Lasso(alpha = 0.1, random_state = 42)
model.fit(X_train_scaled, Y_train)


#evaluating the model's performance
Y_pred = model.predict(X_test_scaled)
mse = mean_squared_error(Y_test, Y_pred)
print(f"Mean squared error on testing set: {mse}")

mpe = np.mean((Y_pred - Y_test) / Y_test) * 100
print(f"Mean percentage error on testing set: {mpe} %")

Mean squared error on testing set: 98.09725887178499
Mean percentage error on testing set: 13.583652406728438 %


### L2 Regularization


In [159]:
#extract the explanatory and response variable
#axis = 1 indicates that we are droping a column
#axis = 1 indicates that we are droping from row
X = df3.drop('Strength', axis = 1)
Y = df3['Strength']

#split into train and test. use 80-20 rule
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2, random_state = 42)

In [160]:
# from sklearn.preprocessing import StandardScaler

# #instance
# scaler = StandardScaler()
# X_train_scaled = scaler.fit_transform(X_train)
# X_test_scaled = scaler.fit_transform(X_test)

# #instance of the model and training
# model = SGDRegressor(loss='squared_error', alpha=0.00001, max_iter=1000, tol=1e-3, random_state=42, verbose = 1)
# model.fit(X_train_scaled, Y_train)


#instance of the model and training
model = Ridge(alpha = 1, random_state = 42)
model.fit(X_train, Y_train)


Ridge(alpha=1, random_state=42)

In [161]:
#evaluating the model's performance
Y_pred = model.predict(X_test)
mse = mean_squared_error(Y_test, Y_pred)
print(f"Mean squared error on testing set: {mse}")

mpe = np.mean((Y_pred - Y_test) / Y_test) * 100
print(f"Mean percentage error on testing set: {mpe} %")

Mean squared error on testing set: 95.97089554036008
Mean percentage error on testing set: 12.805407824372592 %


In [162]:
from sklearn.preprocessing import StandardScaler

#instance
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.fit_transform(X_test)


#instance of the model and training
model = Lasso(alpha = 0.1, random_state = 42)
model.fit(X_train_scaled, Y_train)


#evaluating the model's performance
Y_pred = model.predict(X_test_scaled)
mse = mean_squared_error(Y_test, Y_pred)
print(f"Mean squared error on testing set: {mse}")

mpe = np.mean((Y_pred - Y_test) / Y_test) * 100
print(f"Mean percentage error on testing set: {mpe} %")

Mean squared error on testing set: 98.09725887178499
Mean percentage error on testing set: 13.583652406728438 %


### L2 regularization using Gradeint Descent Algorithm


In [163]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split

# Define your compute_gradient_linear_reg function here

def compute_gradient_linear_reg(X, y, w, b, lambda_):
    """
    Computes the gradient for linear regression with L2 regularization
    
    Args:
      X (ndarray (m,n)): Data, m examples with n features
      y (ndarray (m,)): target values
      w (ndarray (n,)): model parameters  
      b (scalar)      : model parameter
      lambda_ (scalar): Controls amount of regularization
      
    Returns:
      dj_dw (ndarray (n,)): The gradient of the cost w.r.t. the parameters w. 
      dj_db (scalar):       The gradient of the cost w.r.t. the parameter b. 
    """
    m, n = X.shape           #(number of examples, number of features)
    dj_dw = np.zeros((n,))
    dj_db = 0.

    for i in range(m):                             
        err = (np.dot(X[i], w) + b) - y[i]                 
        for j in range(n):                         
            dj_dw[j] += err * X[i, j]               
        dj_db += err                        
    dj_dw /= m                                
    dj_db /= m   
    
    for j in range(n):
        dj_dw[j] += (lambda_ / m) * w[j]

    return dj_db, dj_dw


# # Assume df3 is your dataframe
# df3 = pd.DataFrame({
#     'Feature1': np.random.rand(100),
#     'Feature2': np.random.rand(100),
#     'Feature3': np.random.rand(100),
#     'Strength': np.random.rand(100)
# })

# Separate X and Y
X = df3.drop('Strength', axis=1).values  # Convert to numpy array
Y = df3['Strength'].values  # Convert to numpy array

# Split into train and test sets
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=42)

# Example usage of compute_gradient_linear_reg function
np.random.seed(1)
w_tmp = np.random.rand(X_train.shape[1])  # Initialize weights
b_tmp = 0.5  # Initialize bias
lambda_tmp = 1  # Regularization parameter

# Compute gradient
dj_db_tmp, dj_dw_tmp = compute_gradient_linear_reg(X_train, Y_train, w_tmp, b_tmp, lambda_tmp)

# Print results
print(f"dj_db: {dj_db_tmp}")
print(f"Regularized dj_dw:\n{dj_dw_tmp.tolist()}")


dj_db: 441.3975910478721
Regularized dj_dw:
[126754.37123016847, 36353.3900446637, 20968.843446967097, 80468.21049186478, 2692.4045960339795, 428422.5208231668, 339548.4656900214, 20662.484731241344]


In [164]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split

def compute_gradient_linear_reg(X, y, w, b, lambda_):
    """
    Computes the gradient for linear regression with L2 regularization
    
    Args:
      X (ndarray (m,n)): Data, m examples with n features
      y (ndarray (m,)): target values
      w (ndarray (n,)): model parameters  
      b (scalar)      : model parameter
      lambda_ (scalar): Controls amount of regularization
      
    Returns:
      dj_dw (ndarray (n,)): The gradient of the cost w.r.t. the parameters w. 
      dj_db (scalar):       The gradient of the cost w.r.t. the parameter b. 
      mse (float):          Mean Squared Error for current parameters
    """
    m, n = X.shape           #(number of examples, number of features)
    dj_dw = np.zeros((n,))
    dj_db = 0.
    mse = 0.

    for i in range(m):                             
        err = (np.dot(X[i], w) + b) - y[i]                 
        for j in range(n):                         
            dj_dw[j] += err * X[i, j]               
        dj_db += err
        
        # Compute MSE
        mse += err ** 2

    dj_dw /= m                                
    dj_db /= m
    mse /= m
    
    # Add regularization to gradients
    for j in range(n):
        dj_dw[j] += (lambda_ / m) * w[j]

    return dj_db, dj_dw, mse



# Separate X and Y
X = df3.drop('Strength', axis=1).values  # Convert to numpy array
Y = df3['Strength'].values  # Convert to numpy array

# Split into train and test sets
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=42)

# Example usage of compute_gradient_linear_reg function
np.random.seed(1)
w_tmp = np.random.rand(X_train.shape[1])  # Initialize weights
b_tmp = 0.5  # Initialize bias
lambda_tmp = 0.7  # Regularization parameter
learning_rate = 0.01  # Learning rate for gradient descent
n_iterations = 1000  # Number of iterations

# Perform gradient descent
for iteration in range(n_iterations):
    dj_db_tmp, dj_dw_tmp, mse_tmp = compute_gradient_linear_reg(X_train, Y_train, w_tmp, b_tmp, lambda_tmp)
    
    # Update weights and bias using gradient descent
    w_tmp -= learning_rate * dj_dw_tmp
    b_tmp -= learning_rate * dj_db_tmp
    
    # Display MSE for each iteration
    if iteration % 100 == 0:
        print(f"Iteration {iteration}: MSE = {mse_tmp}")

# Optionally, print final weights and bias
print("\nFinal weights:")
print(w_tmp)
print("Final bias:")
print(b_tmp)


Iteration 0: MSE = 197824.06836992712


/tmp/ipykernel_33/3513319929.py:33: RuntimeWarning: overflow encountered in scalar power
  mse += err ** 2
/tmp/ipykernel_33/3513319929.py:29: RuntimeWarning: overflow encountered in scalar add
  dj_dw[j] += err * X[i, j]
/tmp/ipykernel_33/3513319929.py:29: RuntimeWarning: invalid value encountered in scalar multiply
  dj_dw[j] += err * X[i, j]
/tmp/ipykernel_33/3513319929.py:67: RuntimeWarning: invalid value encountered in subtract
  w_tmp -= learning_rate * dj_dw_tmp


Iteration 100: MSE = nan
Iteration 200: MSE = nan
Iteration 300: MSE = nan
Iteration 400: MSE = nan
Iteration 500: MSE = nan
Iteration 600: MSE = nan
Iteration 700: MSE = nan
Iteration 800: MSE = nan
Iteration 900: MSE = nan

Final weights:
[nan nan nan nan nan nan nan nan]
Final bias:
nan
